### DISCUSSION TITLE
state.changed in v5

### DISCUSSION THREAD

**Discussion by adithep** [statelyai/xstate#4412] [Q&A] [answered]
In v4 we can detect if the transition is successful like so: 

https://xstate.js.org/docs/guides/states.html#state-changed

How do I know if the event I sent is successful in v5? Is it possible currently? If not I think it will be a very useful feature.

**Comment by davidkpiano** [comment, marked as answer]
In XState v5, it is a simple referential equality check:

```diff
let prevSnapshot;

actor.subscribe(snapshot => {
- snapshot.changed;
+ snapshot === prevSnapshot; // false if changed

  prevSnapshot = snapshot;
});
```

cc. @Andarist

**Comment by thirtified** [comment]
It really feels as if this should be a framework feature. Feels a bit weird having to manually keep track of the snapshot state of a state machine 😉.
That is, I'd very much appreciate having such a functionality. And maybe even an option to only trigger subscription events when the snapshot has changed?

**Reply by Andarist** [reply]
> And maybe even an option to only trigger subscription events when the snapshot has changed?

I'm in favor of this. IIRC @davidkpiano had some reservations about it though

**Reply by davidkpiano** [reply]
What would this API (option to subscribe only to changes) look like?

**Reply by Andarist** [reply]
No changes at all. It would just change the frequency of the notifications. If there was no change of the snapshot then the user doesn't have to be notified - they already received that snapshot.

**Reply by thirtified** [reply]
I understand that there wouldn't need to be changes to the snapshots and overall logic, but how would one indicate that a subscriber should only be notified on changes? Maybe some way of providing subscription options, e.g.
```js
 actor.subscribe(observer, {onlyWhenChanged: true});
```
I guess one could in principle take this a step further in terms of generalization and introduce something like interceptors/ middleware/conditions for observers – something that gets invoked before any subscriber and gets to decide the notification behaviors. xstate could come with a set of standard behaviors such as the "onlyWhenChanged" one and allow for easy extension.
```js
 actor.subscribe(observer, {conditions: [onlyWhenChanged]});
```
But maybe that's a bit over the top...

**Reply by Andarist** [reply]
I don't think we want to make this configurable at all. It's not that hard to use some extra helpers to achieve extra behaviors on top of this. That's what you have `distinctUntilChanged` in RxJS for etc - it's a pretty common pattern to have some "operators" for notifications. This can already be implemented today.

I'm just saying that perhaps this could be the default behavior. I don't see huge value in being notified with the same repetitive information and if that could already be handled by XState internally for me, I think it would be for the better.

**Reply by davidkpiano** [reply]
Now that we have the inspect API, I agree that this would be a good default behavior. However, it would be a breaking change in XState v5.

**Reply by yllan** [reply]
Maybe not a xstate problem but a typescript one: I'm trying to implement an utility function like this:

```javascript
const onlyChangeWrapper = (actor) => {
  let prevSnapshot = undefined
  const subscribers = {}

  actor.subscribe((snapshot) => {
    if (snapshot !== prevSnapshot) {
      for (const fn of Object.values(subscribers)) 
        fn(snapshot)
    }
  })

  const subscribe = (fn) => {
    const key = uuid()
    subscribers[key] = fn
    return {
      unsubscribe: () => delete subscribers[key]
    }
  }

  return {
    actor,
    subscribe,
  }
}
```

However I don't know how to type this. Can you shed some light on this?

**Reply by zaccrites** [reply]
@yllan I think you can use `ReturnType<typeof actor.getSnapshot>`

**Reply by yllan** [reply]
@zaccrites Thanks for pointing out that for me. What about the type of actor?

**Reply by thirtified** [reply]
> @zaccrites Thanks for pointing out that for me. What about the type of actor?

Easiest is probably `Actor<any>` as you likely won't care for the underlying state machine. If for some reason you do need to type the snapshots you may want to employ generics and `ActorRef`.
